extensions-core/s3-extensions/src/main/java/org/apache/druid/storage/s3/S3DataSegmentPuller.java

```diff
@@ -84,7 +81,7 @@ public S3DataSegmentPuller(ServerSideEncryptingAmazonS3 s3Client)
     try {
       org.apache.commons.io.FileUtils.forceMkdir(outDir);
 
-      final URI uri = URI.create(StringUtils.format("s3://%s/%s", s3Coords.bucket, s3Coords.path));
+      final URI uri = URI.create(StringUtils.format("s3://%s/%s", s3Coords.getBucket(), s3Coords.getPath()));
```

Comment:
Is the problem that described in https://github.com/apache/incubator-druid/pull/8903/files/a4f6ae9ae2f81381d865e87ec5e1219d275f299c..7125e3e94bd468bc82c95fad68538890a23c69ee#r348869424 possible here when the `URI` created here gets passed to the `CloudObjectLocation` constructor?

Is there a test to check the handling of tricky characters?

Reply:
There is not such a test afaik, I guess can look into this, or maybe as a follow-up since this isn't really new code and sort of feels like the scope of this PR is creeping

Reply:
IMHO we should always leave code better than we found it. Small bugs like this are not worth putting into an issue, and will likely never get worked on, but some poor soul somewhere on the interwebs will run into it and bang their head against it.

Reply:
Fixing bugs is not scope creep...

Reply:
This is not a regression and so doesn't have to be fixed in this PR. It's up to the author in this case. 

> IMHO we should always leave code better than we found it. Small bugs like this are not worth putting into an issue, and will likely never get worked on, but some poor soul somewhere on the interwebs will run into it and bang their head against it.

I don't think this will be happening for this bug. This bug is pretty critical and should be fixed as soon as possible.

Reply:
I went ahead and fixed this for most of S3 by refactoring to use `CloudObjectLocation` to ensure URI handling is good, though the footprint of this PR has grown a lot, which is what I was worried about. In some sense, this is sort of related to the work done in #6761. I have opened #8941 to finish the remaining issues.